src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs

```diff
@@ -372,5 +372,67 @@ public async Task Listener_AwaitsConnection_ListenerSurvivesGC()
                     await connection.DisposeAsync();
                 }).WaitAsync(TimeSpan.FromSeconds(5)));
         }
+
+        [Fact]
+        public async Task Listener_AlpnChange() {
+            var listenerOptions = new QuicListenerOptions()
+            {
+                ListenEndPoint = new IPEndPoint(IPAddress.Loopback, 0),
+                ApplicationProtocols = new List<SslApplicationProtocol>()
+                {
+                    new SslApplicationProtocol("foo"),
+                    new SslApplicationProtocol("bar"),
+                    new SslApplicationProtocol("test"),
+                },
+                ConnectionOptionsCallback = (_, _, _) =>
+                {
+                    // Narrowing down alpn list to more specific.
+                    var options = CreateQuicServerOptions();
+                    options.ServerAuthenticationOptions.ApplicationProtocols = new()
+                    {
+                        new SslApplicationProtocol("bar"),
+                        new SslApplicationProtocol("test"),
+                    };
+                    return ValueTask.FromResult(options);
+                }
+            };
+            await using QuicListener listener = await CreateQuicListener(listenerOptions);
+
+            // Successful connection with bar ALPN
+            QuicClientConnectionOptions clientOptions1 = CreateQuicClientOptions(listener.LocalEndPoint);
+            clientOptions1.ClientAuthenticationOptions.ApplicationProtocols = new()
+            {
+                new SslApplicationProtocol("foo"),
+                new SslApplicationProtocol("bar"),
+            };
+            ValueTask<QuicConnection> connectTask1 = CreateQuicConnection(clientOptions1);
+            await using QuicConnection serverConnection1 = await listener.AcceptConnectionAsync().AsTask().WaitAsync(TimeSpan.FromSeconds(30));
```

Comment:
I know, I know, that the tests above also have `.WaitAsync(TimeSpan.FromSeconds(30));` all over them, but I propose to maybe do it via test-wide CancellationToken for a bit less code (and magical 30 seconds) and a bit more clarity? We should have a defined constant PassingTestTimeout, so I propose

```c#
public async Task Listener_AlpnChange() {
    using CancellationTokenSource testTimeoutCts = new CancellationTokenSource(PassingTestTimeout);
    CancellationToken timeoutToken = testTimeoutCts.Token;

    // ...

    await using QuicConnection serverConnection1 = await 
listener.AcceptConnectionAsync().AsTask().WaitAsync(timeoutToken);
```

WDYT? /cc @ManickaP @wfurt

Reply:
Or just pass the `PassingTestTimeout` instead of the magic constant?
But in general I like Natalia's suggestion.